File: sicp_1_21.rb
# find smallest divisor of an integer
# adapted from SICP section 1.2

def square(n)
  n * n
end  

def smallest_divisor(n)
  find_divisor(n,2)
end

def find_divisor(n,test_divisor)
  case
  when square(test_divisor) > n
    n
  when divides(test_divisor,n)
    test_divisor
  else
    find_divisor(n,(test_divisor+1))
  end
end

def divides(a,b)
  b % a == 0
end

def prime(n)
  smallest_divisor(n) == n
end

[199,1999,19999].each do |number|
  puts "#{number} / smallest divisor: #{smallest_divisor(number)}"
end  

File: sicp_1_21.c
// sicp 1_21 in c

#include <stdio.h>

int square(int number)
{
    return number * number;
}

int divides(int a, int b)
{
    int booool = b % a;
    if (booool == 0) {
        return 0;
    } else {
        return 1;
    }
}

int findDivisor(int number, int testDivisor)
{
    if (square(testDivisor) > number) {
        return number;
    } else if (divides(testDivisor, number) == 0) {
        return testDivisor;
    } else {
        int varTwo = testDivisor + 1;
        int value = findDivisor(number, varTwo);
        return value;
    }
}

int smallestDivisor(int x)
{
    int value = findDivisor(x, 2);
    return value;
}

int prime(int n)
{
    int value = smallestDivisor(n);
    if (value == n) {
        return 0;
    } else {
        return 1;
    }
}

int main(int argc, const char * argv[]) {
    int a = smallestDivisor(199);
    int b = smallestDivisor(1999);
    int c = smallestDivisor(19999);
    printf("%d, %d, %d\n", a, b, c);
    return 0;
}
